packages/driver/src/cross-origin/patches/cookies.js

```diff
@@ -13,43 +56,59 @@
 // - On an interval, get the browser's cookies for the given domain, so that
 //   updates to the cookie jar (via http requests, cy.setCookie, etc) are
 //   reflected in the document.cookie value.
-export const patchDocumentCookie = (Cypress, window) => {
-  const setAutomationCookie = (toughCookie) => {
-    const { superDomain } = Cypress.Location.create(window.location.href)
-    const automationCookie = Cypress.Cookies.toughCookieToAutomationCookie(toughCookie, superDomain)
+export const handleDocumentCookie = (Cypress) => {
+  // if cookies were sent before the spec bridge was created, they'll come
+  // through as state('crossOriginCookies'), so use them then reset them
+  const cookies = Cypress.state('crossOriginCookies')
+  let documentCookieValue = mergeCookies(document.cookie, cookies)
+
+  Cypress.state('crossOriginCookies', [])
 
-    Cypress.automation('set:cookie', automationCookie)
-    .catch(() => {
+  const patch = (window) => {
+    // Cookies added to the server-side cookie jar are optimistically
+    // added here so that if a cross-origin request sets cookies, they're
+    // available via document.cookie synchronously on page load
+    const setAutomationCookie = (toughCookie) => {
+      const { superDomain } = Cypress.Location.create(window.location.href)
+      const automationCookie = Cypress.Cookies.toughCookieToAutomationCookie(toughCookie, superDomain)
+
+      Cypress.automation('set:cookie', automationCookie)
+      .catch(() => {
       // unlikely there will be errors, but ignore them in any case, since
       // they're not user-actionable
-    })
-  }
+      })
+    }
 
-  let documentCookieValue = ''
+    Object.defineProperty(window.document, 'cookie', {
+      get () {
+        return documentCookieValue
+      },
 
-  Object.defineProperty(window.document, 'cookie', {
-    get () {
-      return documentCookieValue
-    },
+      set (newValue) {
+        const cookie = Cypress.Cookies.parse(newValue)
 
-    set (newValue) {
-      const cookie = Cypress.Cookies.parse(newValue)
+        // If cookie is undefined, it was invalid and couldn't be parsed
+        if (!cookie) return documentCookieValue
 
-      // If cookie is undefined, it was invalid and couldn't be parsed
-      if (!cookie) return documentCookieValue
+        documentCookieValue = mergeCookies(documentCookieValue, [{
```

Comment:
I don't think this currently preserves duplicates, such as 

```
const merged = mergeCookies('foo=bar; foo=bar', [{
  name: 'foo',
  value: 'baz',
}])

// merged is 'foo=baz', where I would expect something like 'foo=bar; foo=baz'
```

I played around with some cookies in chrome and firefox to spoof the following
![](https://user-images.githubusercontent.com/3980464/189220674-05bd99a7-b624-4afb-b29b-c195773bef3a.png)
![](https://user-images.githubusercontent.com/3980464/189220778-a27d8002-7061-4017-8328-572f010dd20d.png)
![](https://user-images.githubusercontent.com/3980464/189220755-a4021b7a-19cf-488f-80d9-60b29bd33adb.png)


I think we need more to go off of to set the cookie here properly, such as cookie domain and path, which it looks like we get from the `get:cookies` automation response.

![Screen Shot 2022-09-08 at 4 25 46 PM](https://user-images.githubusercontent.com/3980464/189220555-c5f1798d-9fee-4bae-a52f-b8eb78da47c2.png)

I wonder if we can take the cookies from the `setInterval` , i.e `Cypress.automation('get:cookies', { domain })` and store those within the `handleDocumentCookie` closure similar to how we store the `documentCookieValue`. Then, when a user gets `document.cookie`, we just stringify the cookies from the automation response. When a user sets a cookie, we can take the cookies from the variable that has the reference to the `get:cookies` stored variable and make sure we overwrite the correct cookie. The default should be the full `hostname` with path `/`, which I think `tough-cookie` handles but I can't remember off the top of my head.

Reply:
Fixed and [added a test](https://github.com/cypress-io/cypress/pull/23643/files#diff-6bf45b7f9903569c8f01830b58bfeac19264dc6b4eee767155d1b6484492b25fR854). Now use a `tough-cookie` cookie jar to track the cookies, so it applies the correct logic for handling duplicate cookie keys.